uploadNeuroDB/file_insertion.pl

```diff
@@ -0,0 +1,509 @@
+MESSAGE
+    # write error message in the log file
+    $utility->writeErrorLog( $message, $INVALID_SCANNER_ID, $log_file );
+    # insert error message into notification spool table
+    $notifier->spool(
+        'file insertion'   , $message,   0,
+        'file_insertion.pl', $upload_id, 'Y',
+        'N'
+    );
+    exit $INVALID_SCANNER_ID; ##TODO 1: call the exit code from ExitCodes.pm
+}
+
+
+
+
+##### Verify that the upload ID refers to a valid upload ID
+(my $query = <<QUERY) =~ s/\n/ /gm;
+SELECT UploadID FROM mri_upload WHERE UploadID=?
+QUERY
+my $sth = $dbh->prepare($query);
+$sth->execute($upload_id);
+unless ($sth->rows > 0) {
+    # if no row returned, exits with message that did not find this upload ID
+    $message = <<MESSAGE;
+    ERROR: Invalid UploadID $upload_id.\n\n
+MESSAGE
+    # write error message in the log file
+    $utility->writeErrorLog( $message, $INVALID_UPLOAD_ID, $log_file );
+    # insert error message into notification spool table
+    $notifier->spool(
+        'file insertion'   , $message,   0,
+        'file_insertion.pl', $upload_id, 'Y',
+        'N'
+    );
+    exit $INVALID_UPLOAD_ID; ##TODO 1: call the exit code from ExitCodes.pm
+}
+
+
+
+
+###### Determine the acquisition protocol ID of the file based on scan type
+
+# verify that an acquisition protocol ID exists for $scan_type
+my $acqProtocolID = NeuroDB::MRI::scan_type_text_to_id($scan_type, \$dbh);
+if ($acqProtocolID =~ /unknown/){
+    $message = <<MESSAGE;
+    ERROR: no acquisition protocol ID found for $scan_type.\n\n
+MESSAGE
+    # write error message in the log file
+    $utility->writeErrorLog( $message, $PROTOCOL_ID_NOT_FOUND, $log_file );
+    # insert error message into notification spool table
+    $notifier->spool(
+        'file insertion',    $message,   0,
+        'file_insertion.pl', $upload_id, 'Y',
+        'N'
+    );
+    exit $PROTOCOL_ID_NOT_FOUND; ##TODO 1: call the exit code from ExitCodes.pm
+} else {
+    $message = <<MESSAGE;
+    Found acquisition protocol ID $acqProtocolID for $scan_type.\n\n
+MESSAGE
+    $notifier->spool(
+        'file insertion',    $message,   0,
+        'file_insertion.pl', $upload_id, 'Y',
+        'Y'
+    )
+}
+
+
+
+
+###### Create and Load File object
+
+# create File object
+my $file = NeuroDB::File->new(\$dbh);
+
+# file type determined here as long as the it exists in ImagingFileTypes table
+$file->loadFileFromDisk($file_path);
+
+
+
+
+###### Determine file basename and directory name
+
+# Get the file name and directory name of $file_path
+my ($file_name, $dir_name) = fileparse($file_path);
+
+
+
+
+###### Determine the metadata to be stored in parameter_file
+
+##TODO: read a simply ASCII or JSON file to grep the metadata (would be
+# created by the overall insertion scripts like PET_loader etc...)
+
+
+
+
+###### Determine candidate information
+
+# create a hash similar to tarchiveInfo so that can use Utility routines to
+# determine the candidate information
+my %info;
```

Comment:
Could also be done more concisely as

    my %info = (
        SourceLocation => undef,
        PatientName    => ($patient_name // $file_name),
        PatientID      => ($patient_name // $file_name)
    );

Reply:
👍 Love it!!